### PR TITLE
Glue for EcDh, Hash and Hmac

### PR DESCRIPTION
**Pull request by nnmkhang**
Added: 
- Glue for Hmac
- Glue for Hash
- Glue for EcDh
- Comments for each file
- main.rs as a temp testing file
- mod.rs ( WIP don't review)

### CODE REVIEW

File: src/ecdh.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+///
+/// [`start()`] creates a new symcrypt_sys::EcDh struct and subsequently, a new KeyExchange struct.
+/// symcrypt_sys::symcrypt_init() must be called before any EcDh functions. This call is blocking and any subsequent calls to
+/// symcrypt_init() will be no-ops. Specific errors from symcrypt_sys::ecdh will not be tracked and not propagated up to the caller,
+/// [`GetRandomFailed`] will be returned instead.
+///
+/// [`name()`] returns the NamedGroup of the current KeyExchange group.
+impl SupportedKxGroup for KxGroup {
+    fn start(&self) -> Result<Box<(dyn ActiveKeyExchange + 'static)>, GetRandomFailed> {
+        symcrypt_init(); // Calling before symcrypt_sys::ecdh::EcDh::new()
```
Comment (PhilHallin) [inline src/ecdh.rs]: > symcrypt_init() [](http://example.com/codeflow?start=7&length=16)

Seems like this should be called in the EcDh::new function? #Closed

File: Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "rustls-symcrypt"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+rustls = { git = "https://github.com/rustls/rustls.git" }
+rust-symcrypt = {version = "0.1", path = "C:\\Code\\rust-symcrypt-sys\\rust-symcrypt"}
```
Comment (eholk) [inline Cargo.toml:10]: I'd recommend not checking in paths to your own machine. Relative paths are better, and it's best when the path points to the same repo.

If these two crates are likely to have a lot of co-development going forward, I'd actually recommend moving them into the same repo and using cargo workspaces. This is a common pattern for projects that release a number of related crates. For example, see the way Tokio's repo is laid out: https://github.com/tokio-rs/tokio

File: src/ecdh.rs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+/// KeyExchange is a struct that defines the state for EcDh operations
+///
+/// the [`state'] field is tied to the ['EcDh'] struct from symcrypt_sys.
+///
+/// the ['name'] and ['curve_type'] provide access to the rustls::NamedGroup
+/// and symcrypt_sys::ecurve::CurveType respectively
+///
+/// ['pub_key'] is a Vec<u8> that represents the public key that is tied to the EcDh state.
+/// The private_key is not exposed.
```
Comment (eholk) [inline src/ecdh.rs]:
```suggestion
/// KeyExchange is a struct that defines the state for EcDh operations
///
/// the [`state`] field is tied to the [`EcDh`] struct from symcrypt_sys.
///
/// the [`name`] and [`curve_type`] provide access to the rustls::NamedGroup
/// and symcrypt_sys::ecurve::CurveType respectively
///
/// [`pub_key`] is a Vec<u8> that represents the public key that is tied to the EcDh state.
/// The private_key is not exposed.
```

File: src/ecdh.rs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+pub static SECP256R1: &dyn SupportedKxGroup = &KxGroup {
+    name: NamedGroup::secp384r1,
+    curve_type: eckey::CurveType::NistP384,
+};
+
+pub static SECP384R1: &dyn SupportedKxGroup = &KxGroup {
+    name: NamedGroup::secp384r1,
+    curve_type: eckey::CurveType::NistP256,
+};
```
Comment (eholk) [inline src/ecdh.rs]: Instead of using `&dyn SupportedKxGroup` here, you can probably use just `KxGroup` or maybe `&KxGroup`.

Also, does `pub const` instead of `pub static` work here?

File: src/ecdh.rs [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+/// Impl for the trait SupportedKxGroup
+///
+/// [`start()`] creates a new symcrypt_sys::EcDh struct and subsequently, a new KeyExchange struct.
+/// symcrypt_sys::symcrypt_init() must be called before any EcDh functions. This call is blocking and any subsequent calls to
```
Comment (eholk) [inline src/ecdh.rs]: What happens if `symcrypt_init()` is not called? Do you get something like a runtime panic or abort, or does it cause undefined behavior?

File: src/hash.rs [hunk 6]
```diff
@@ -0,0 +1,120 @@
+//! Hash functions. For further documentation please refer to rust_symcrypt::hash
+//!
+use rust_symcrypt::hash::{sha256, sha384, HashState, Sha256State, Sha384State};
+use rustls::crypto::hash::{Context, Hash, HashAlgorithm, Output};
+
+/// ShaXXX is a struct that represents either a Sha256 or Sha384 hash Algorithm
```
Comment (eholk) [inline src/hash.rs:6]: It'd be good to run this file through rustdoc to make sure the docs look right. I think you may want all these to be prefixed with `//!` to make the docs apply for the whole file rather than the `Sha256` struct that comes right after it.

File: src/hash.rs [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+/// Structs related to Sha256
+pub struct Sha256;
+pub struct Sha256Context(Sha256State);
+
+/// Structs related to Sha384
+pub struct Sha384;
+struct Sha384Context(Sha384State);
+
+/// Impl's for Sha256 related traits
+impl Hash for Sha256 {
```
Comment (eholk) [inline src/hash.rs:47]: Since you only have two types (`Sha256` and `Sha384`) this suggestion probably isn't worth it, but you could save some typing here if you wrote a macro to generate the `Hash` and `Context` impls.

File: src/main.rs [hunk 8]
```diff
@@ -0,0 +1,61 @@
+use rustls::crypto::hmac::Hmac;
+use rustls_symcrypt::ecdh::*;
+/// TEMP FILE FOR TESTING DO NOT REVIEW
```
Comment (eholk) [inline src/main.rs]: If you wanted to make this file a little less temporary, you could make it an example: https://doc.rust-lang.org/cargo/reference/cargo-targets.html#examples

That could be helpful for showing people how to use the APIs.

File: src/mod.rs [hunk 9]
```diff
@@ -0,0 +1,42 @@
+//! WIP DO NOT REVIEW
```
Comment (eholk) [inline src/mod.rs]: Are you planning to finish this before merging this PR, or does it make sense to pull it into a separate PR?

File: src/ecdh.rs (hunk 5 above)
Comment (nnmkhang) [reply src/ecdh.rs]: if `SymCryptInit()` is not called before, there will be a panic. I've moved the `Init()` over to the wrapper side for when we initialize the curves, I will remove this comment

File: src/mod.rs (hunk 9 above)
Comment (nnmkhang) [reply src/mod.rs]: Sorry, I will remove this file it was meant for a different PR

File: src/main.rs (hunk 8 above)
Comment (nnmkhang) [reply src/main.rs]: This file is just a local file used for testing will remove from the PR

File: src/hash.rs (hunk 6 above)
Comment (nnmkhang) [reply src/hash.rs:6]: Ran `cargo doc` , looks like theres a lot of things to be fixed between this repo and rust-symcrypt. I will file an issue against both and fix them in one go

File: Cargo.toml (hunk 2 above)
Comment (nnmkhang) [reply Cargo.toml:10]: Im not sure that these crates will have much co-development in the future. This crate will just have dependency on rust-symcrypt and should not change once the traits are completed. I will file the cargo workspace as an issue and look to merge the two together

File: src/ecdh.rs (hunk 4 above)
Comment (nnmkhang) [reply src/ecdh.rs]: [SupportedKxGroup](https://github.com/nnmkhang/rustls-symcrypt/pull/2/files#diff-2ada98d540e88906bfe287caa999ddd6afa38d46fc8733ce014857d4cdd5801eR63:~:text=impl-,SupportedKxGroup,-for%20KxGroup%20%7B) is needed for the impl defined by rustls, I think that `SupportedKxGroup` is needed here?

Comment (eholk) [reply src/ecdh.rs]: As long as you have `impl SupportedKxGroup for KxGroup` then the compiler should be able to figure out that `KxGroup` meets the `SupportedKxGroup` requirement. I think even if rustls needs a `&dyn SupportedKxGroup`, the compiler should be able to automatically coerce a `&KxGroup` into a `&dyn SupportedKxGroup`. I'm not 100% sure though, there may be other requirements that I'm missing.

**Review by PhilHallin — APPROVED** [review]
:shipit:
